add BGP route injection

polynomialspace opened the pull request:
- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.

<!-- Description of change -->

This allows the user the ability to advertise routes where the LoadBalancer pattern is not desired.  Additionally, allows the user to advertise ClusterIPs in cases where non-kubernetes workloads must access the cluster.
```release-note
Allow injecting routes to BGP control plane announcements.
```

qmonnet commented on pkg/bgpv1/gobgp/reconcile_test.go:
```diff
@@ -473,3 +473,129 @@ func TestExportPodCIDRReconciler(t *testing.T) {
+				var seen bool
+				for _, advrt := range testSC.InjectedRoutes {
+					if advrt.Net.String() == parsed.String() {
+						seen = true
```
Nit: break out of the loop?

polynomialspace replied:
artifact from copying other tests, corrected both instances only in this test

qmonnet commented on pkg/bgpv1/gobgp/reconcile_test.go:
```diff
@@ -473,3 +473,129 @@ func TestExportPodCIDRReconciler(t *testing.T) {
+					}
+				}
+				if !seen {
+					t.Fatalf("unwated advert %+v", advrt)
```
```suggestion
					t.Fatalf("unwanted advert %+v", advrt)
```

polynomialspace replied:
another copypaste artifact, fixed for TestExportPodCIDRReconciler as well

qmonnet commented on Documentation/network/bgp-control-plane.rst:
```diff
@@ -73,6 +75,9 @@ Fields
            neighbors[*].peerAddress: The address of the peer neighbor
            neighbors[*].peerASN: The ASN of the peer
 
+       virtualRouters[*].injectedRoutes.prefixes: List of additional routes to be "injected" into announcements.
```
Would be nice to have a few more details on injected routes in the doc (thinking of your comment on skipping the LB, or advertising ClusterIPs, from your commit description).

qmonnet reviewed (changes requested):
Thanks for this work! Looks good overall, although I'm mildly familiar with BGP. Please find a few minor comments below.

qmonnet reviewed (approved):
Thanks!

sayboras commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -111,4 +111,17 @@ type CiliumBGPVirtualRouter struct {
+	// InjectedRoutes is a list of BGP Routes to be announced for this virtual router
+	//
+	// +kubebuilder:validation:Optional
+	InjectedRoutes CiliumBGPInjectedRoute `json:"injectedRoutes"`
```
as this one is optional, any reason for not using a pointer instead of a value here? Also omitempty will be useful as well.

polynomialspace replied:
Slipped my mind, updated

Additionally checked in the api crd I'd apparently forgotten

sayboras reviewed (commented):
I am not familiar with bgp as well, so just review k8s API.

sayboras reviewed (approved):
LGTM for API changes :100:

dylandreimerink reviewed (changes requested):
Sorry that it has taken someone from the BGP SIG this long to take a look at this. First of all, thank you for the work, really appreciate it.

There are however 2 thinks that I would like to see changed/added.
The first is that I would suggest a name change from `injectedRoutes` to `additionalRoutes` since that better explains what we are doing. And changing `prefixes` to `cidrs` to match the existing nomenclature.

The second is that the current CRD doesn't allow for future expansion. We currently only announce IPv4 and IPv6 Unicast routes, but I would like to keep the door open for additional route types to be added(Flowspec, Multicast, ect.). To that end I suggest changing the CRD to allow the following structure:
```yaml
type: CiliumBGPVirtualRouter
metadata:
  name: "router-a"
spec:
  [...]
  additionalRoutes:
  - afi: ipv4
    safi: unicast
    cidrs: 
    - '10.0.10.0/24'
  # To allow for future routes of different afi-safi 's
  #- afi: ipv4
  #  safi: flowspec
  #  rules:
  #  - match:
  #      source: '30.0.0.0/8'
  #    action:
  #      discard: true
```

I have included an example of what a flowspec route might look like in the future. The idea being that we can add additional fields for newer route type once we start supporting them without having to break compatibility.

polynomialspace committed:
bgp: allow user-specified routes for announcement.

This allows the user the ability to advertise routes where the LoadBalancer pattern is not desired.  Additionally, allows the user to advertise ClusterIPs in cases where non-kubernetes workloads must access the cluster.

Signed-off-by: Ash Burn <github@polynomial.space>
